kernel/kernel_drivers.go

```diff
@@ -0,0 +1,208 @@
+		switch node.Type() {
+		case 0, fs.ModeDir:
+			// Create link for regular files or directories
+			if err := os.Symlink(origPath, lpath); err != nil {
+				return err
+			}
+		case fs.ModeSymlink:
+			// Replicate link (it should be relative)
+			lpath := filepath.Join(fwDest, node.Name())
+			dest, err := os.Readlink(origPath)
+			if err != nil {
+				return err
+			}
+			if err := os.Symlink(dest, lpath); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("%q has unexpected file type: %s",
+				node.Name(), node.Type())
+		}
+	}
+
+	return nil
+}
+
+func createModulesSubtree(origin, dest string) error {
+	kversion, err := KernelVersionFromModulesDir(origin)
+	if err != nil {
+		// Bit of a corner case, but maybe possible. Log anyway.
+		// TODO if not found but we have components with kernel modules
+		// we raise an error (but maybe from other place?)
+		logger.Noticef("no modules found in %q", origin)
+		return nil
+	}
+
+	// Although empty we need "lib" because "depmod" always appends
+	// "/lib/modules/<kernel_version>" to the directory passed with option
+	// "-b".
+	modsRoot := filepath.Join(dest, "lib", "modules", kversion)
+	if err := os.MkdirAll(modsRoot, 0755); err != nil {
+		return err
+	}
+
+	// Copy modinfo files from the snap (these might be overwritten if
+	// kernel-modules components are installed).
+	modsGlob := filepath.Join(origin, "modules", kversion, "modules.*")
+	modFiles, err := filepath.Glob(modsGlob)
+	if err != nil {
+		// Should not really happen (only possible error is ErrBadPattern)
+		return err
+	}
+	for _, orig := range modFiles {
+		target := filepath.Join(modsRoot, filepath.Base(orig))
+		if err := osutil.CopyFile(orig, target, osutil.CopyFlagDefault); err != nil {
+			return err
+		}
+	}
+
+	// Symbolic links to early mount kernel snap
+	earlyMntDir := filepath.Join(origin, "modules", kversion)
+	for _, d := range []string{"kernel", "vdso"} {
+		lname := filepath.Join(modsRoot, d)
+		to := filepath.Join(earlyMntDir, d)
+		if err := osSymlink(to, lname); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func driversTreeDir(ksnapName string, rev snap.Revision) string {
+	return filepath.Join(dirs.SnapdStateDir(dirs.GlobalRootDir), "kernel", ksnapName, rev.String())
+}
+
+// RemoveKernelDriversTree cleans-up the writable kernel tree for a given snap
+// and revision.
+func RemoveKernelDriversTree(ksnapName string, rev snap.Revision) (err error) {
+	treeRoot := driversTreeDir(ksnapName, rev)
+	return os.RemoveAll(treeRoot)
+}
+
+// EnsureKernelDriversTree creates a tree in
+// /var/lib/snapd/kernel/<ksnapName>/<rev> with subfolders can be bind-mounted
+// on boot to /usr/lib/{modules,firmware}. This tree contains files from the
+// kernel snap mounted on kernelMount, as well as symlinks to it.
+// TODO this will be extended to consider kernel-modules components.
+func EnsureKernelDriversTree(ksnapName string, rev snap.Revision, kernelMount string) (err error) {
```

Comment:
nitpick: do we need kernelMount?  can't we use snap package functions to compute  it from name and revno?

Reply:
In principle I want to use here the snap mounted in `/run/mnt/kernel-snap/<name>/<rev>` (it could be the one under `/snap/...`, but I think it is more consistent this way) - and PlaceInfo implementation assumes /snap